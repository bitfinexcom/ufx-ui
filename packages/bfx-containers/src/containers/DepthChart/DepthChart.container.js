/* eslint-disable react/jsx-props-no-spreading */
import { DepthChart, PROP_DEFAULT_CCYS } from '@ufx-ui/core'
import React, { memo } from 'react'
import { useSelector } from 'react-redux'

import useCommonBfxData from '../../hooks/useCommonBfxData'
import { set } from '../../redux/actions/UI.actions'
import { DEPTH_CHART_ZOOM } from '../../redux/constants/UI.constants'
import {
  getBookTopAsks,
  getBookTopBids,
  getBookTopPasks,
  getBookTopPbids,
  getBookTopTasks,
  getBookTopTbids,
} from '../../redux/selectors/book-top.selectors'
import {
  getBookAsks,
  getBookBids,
  getBooktAsks,
  getBooktBids,
  getBookpBids,
  getBookpAsks,
  getBookSnapshotReceived,
} from '../../redux/selectors/book.selectors'
import { getUrlPair, getUIDepthChartZoom, getUIBookPrec } from '../../redux/selectors/UI.selectors'
import { depthChart } from '../../var/config'

const DepthChartContainer = ({
  baseCcy,
  quoteCcy,
  parentWidth,
  parentHeight,
  paddingX,
  paddingY,
  ...props
}) => {
  const { symbol, dispatch } = useCommonBfxData(baseCcy, quoteCcy)

  const asks = useSelector(state => getBookAsks(state, symbol))
  const bids = useSelector(state => getBookBids(state, symbol))
  const pAsks = useSelector(state => getBookpAsks(state, symbol))
  const pBids = useSelector(state => getBookpBids(state, symbol))
  const tAsks = useSelector(state => getBooktAsks(state, symbol))
  const tBids = useSelector(state => getBooktBids(state, symbol))

  const asksTop = useSelector(getBookTopAsks)
  const bidsTop = useSelector(getBookTopBids)
  const pAsksTop = useSelector(getBookTopPasks)
  const pBidsTop = useSelector(getBookTopPbids)
  const tAsksTop = useSelector(getBookTopTasks)
  const tBidsTop = useSelector(getBookTopTbids)

  const snapshotReceived = useSelector(state => getBookSnapshotReceived(state, symbol))
  const zoom = useSelector(getUIDepthChartZoom)
  const prec = useSelector(getUIBookPrec)
  const currentPair = useSelector(getUrlPair)

  const updateZoom = (value) => {
    dispatch(set({
      section: DEPTH_CHART_ZOOM,
      value: Math.min(Math.max(depthChart.MIN_ZOOM, zoom + value), depthChart.MAX_ZOOM),
    }))
  }

  return (
    <DepthChart
      theme='light-theme'
      width={parentWidth - paddingX * 2}
      height={parentHeight - paddingY * 2}
      asks={asks}
      bids={bids}
      tAsks={tAsks}
      tBids={tBids}
      pAsks={pAsks}
      pBids={pBids}
      asksTop={asksTop}
      bidsTop={bidsTop}
      pAsksTop={pAsksTop}
      pBidsTop={pBidsTop}
      tAsksTop={tAsksTop}
      tBidsTop={tBidsTop}
      loading={!snapshotReceived}
      zoom={zoom}
      prec={prec}
      updateZoom={updateZoom}
      fontFamily='"Source Sans Pro"'
      currentPair={currentPair}
      {...props}
    />
  )
}

DepthChartContainer.propTypes = PROP_DEFAULT_CCYS.props

DepthChartContainer.defaultProps = PROP_DEFAULT_CCYS.defaults

export default memo(DepthChartContainer)
