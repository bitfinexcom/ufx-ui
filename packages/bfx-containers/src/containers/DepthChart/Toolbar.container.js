import { DepthChartToolbar } from '@ufx-ui/core'
import React, { useCallback, memo } from 'react'
import { useSelector, useDispatch } from 'react-redux'

import { set } from '../../redux/actions/UI.actions'
import { DEPTH_CHART_ZOOM } from '../../redux/constants/UI.constants'
import { getUIDepthChartZoom } from '../../redux/selectors/UI.selectors'
import { depthChart } from '../../var/config'

const DepthChartToolbarContainer = () => {
  const zoom = useSelector(getUIDepthChartZoom)

  const dispatch = useDispatch()
  const updateZoom = useCallback(
    (value) => {
      dispatch(set({
        section: DEPTH_CHART_ZOOM,
        value: Math.min(Math.max(depthChart.MIN_ZOOM, zoom + value), depthChart.MAX_ZOOM),
      }))
    },
    [dispatch, zoom],
  )

  return (
    <DepthChartToolbar
      config={depthChart}
      zoom={zoom}
      updateZoom={updateZoom}
    />
  )
}

export default memo(DepthChartToolbarContainer)
