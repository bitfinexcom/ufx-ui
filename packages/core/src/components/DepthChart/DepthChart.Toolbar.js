import { faSearchMinus, faSearchPlus } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import cx from 'classnames'
import React, { memo } from 'react'
import { useTranslation } from 'react-i18next'

import * as Classes from '../../common/classes'
import withI18nProvider from '../../hoc/withI18nProvider'
import Button from '../ui/Button'

const DepthChartToolbar = ({
  zoom,
  updateZoom,
  config,
}) => {
  const classes = cx(`${Classes.DEPTH_CHART}__toolbar`, Classes.RIGHT_TO_LEFT)
  const { t } = useTranslation('depthchart')

  return (
    <div className={classes}>
      <Button
        title={t('decrease_zoom')}
        onClick={() => updateZoom(10)}
        minimal
        icon='minus'
        disabled={zoom === config.MAX_ZOOM}
      >
        <FontAwesomeIcon icon={faSearchMinus} />
      </Button>
      <Button
        title={t('increase_zoom')}
        onClick={() => updateZoom(-10)}
        minimal
        disabled={zoom === config.MIN_ZOOM}
      >
        <FontAwesomeIcon icon={faSearchPlus} />
      </Button>
    </div>
  )
}

export default withI18nProvider(memo(DepthChartToolbar))
