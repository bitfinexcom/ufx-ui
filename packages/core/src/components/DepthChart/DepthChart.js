import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { memo, useEffect, useRef } from 'react'
import { useTranslation } from 'react-i18next'

import * as Classes from '../../common/classes'
import { getCurrentStyle } from '../../common/themes'
import withI18nProvider from '../../hoc/withI18nProvider'
import { PROP_BOOK_TRADE, PROP_BOOK } from '../Book/Book.props'
import Spinner from '../ui/Spinner'
import DepthChartCanvas from './DepthChart.canvas'

export const DepthChart = ({
  asks,
  bids,
  tAsks,
  tBids,
  pAsks,
  pBids,
  asksTop,
  bidsTop,
  tAsksTop,
  tBidsTop,
  pAsksTop,
  pBidsTop,
  className,
  loading,
  theme,
  width,
  height,
  zoom,
  prec,
  updateZoom,
  fontFamily,
  currentPair,
}) => {
  const canvasRef = useRef(null)
  const chartRef = useRef(null)
  const { t } = useTranslation()
  const classes = cx(Classes.DEPTH_CHART, className)
  const { depthChart: depthChartStyle } = getCurrentStyle(theme)

  useEffect(() => {
    // when the width/height changes (due to resizing) unset charts so that
    // they will be re-instantiated with the new width/height in the other
    // useEffect block. Do the same when disconnected/loading
    chartRef.current = null
  }, [width, height, currentPair, prec])

  useEffect(() => {
    if (!canvasRef.current || !pBidsTop.length || !pAsksTop.length) return

    if (!chartRef.current) {
      chartRef.current = new DepthChartCanvas({
        ref: canvasRef.current,
        bidsColor: depthChartStyle.bidsColor,
        asksColor: depthChartStyle.asksColor,
        bgColor: depthChartStyle.bgColor,
        axisColor: depthChartStyle.axisColor,
        fontColor: depthChartStyle.fontColor,
        highlightColor: depthChartStyle.highlightColor,
        initialZoom: zoom,
        fontFamily,
        onScroll: event => {
          updateZoom((event.deltaY * 0.02))
        },
      })
    }

    chartRef.current.render({
      bids,
      tBids,
      pBids,
      asks,
      tAsks,
      pAsks,
      bidsTop,
      tBidsTop,
      pBidsTop,
      asksTop,
      tAsksTop,
      pAsksTop,
      zoom,
    })
  }, [bids, tBids, pBids, asks, tAsks, pAsks, bidsTop, tBidsTop, pBidsTop, asksTop, tAsksTop, pAsksTop, depthChartStyle, zoom, updateZoom, fontFamily])

  if (loading) {
    return (
      <div className={`${Classes.DEPTH_CHART}__error`}>
        {loading && (
          <Spinner />
        )}
      </div>
    )
  }

  if (!(pBidsTop.length > 2 && pAsksTop.length > 2)) {
    return <div className='side-message'>{t('common:no_data_available')}</div>
  }

  return (
    <div className={classes}>
      <canvas ref={canvasRef} className='canvas' width={width || undefined} height={height || undefined} />
    </div>
  )
}

DepthChart.propTypes = {
  loading: PropTypes.bool,
  asks: PropTypes.objectOf(PropTypes.shape(PROP_BOOK)).isRequired,
  bids: PropTypes.objectOf(PropTypes.shape(PROP_BOOK)).isRequired,
  tAsks: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.shape(PROP_BOOK_TRADE), PropTypes.number]),
  ).isRequired,
  tBids: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.shape(PROP_BOOK_TRADE), PropTypes.number]),
  ).isRequired,
  pAsks: PropTypes.arrayOf(PropTypes.number).isRequired,
  pBids: PropTypes.arrayOf(PropTypes.number).isRequired,
  asksTop: PropTypes.objectOf(PropTypes.shape(PROP_BOOK)).isRequired,
  bidsTop: PropTypes.objectOf(PropTypes.shape(PROP_BOOK)).isRequired,
  tAsksTop: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.shape(PROP_BOOK_TRADE), PropTypes.number]),
  ).isRequired,
  tBidsTop: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.shape(PROP_BOOK_TRADE), PropTypes.number]),
  ).isRequired,
  pAsksTop: PropTypes.arrayOf(PropTypes.number).isRequired,
  pBidsTop: PropTypes.arrayOf(PropTypes.number).isRequired,
  className: PropTypes.string,
  theme: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
  zoom: PropTypes.number,
  prec: PropTypes.number,
  updateZoom: PropTypes.func.isRequired,
  fontFamily: PropTypes.string,
  currentPair: PropTypes.string.isRequired,
}

export const defaultProps = {
  loading: false,
  className: null,
  theme: 'eosfinex-dark-theme',
  width: 600,
  height: 220,
  zoom: 50,
  prec: 1,
  fontFamily: '"Helvetica Neue"',
}

DepthChart.defaultProps = defaultProps

export default withI18nProvider(memo(DepthChart))
