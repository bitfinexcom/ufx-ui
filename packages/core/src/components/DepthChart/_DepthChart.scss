.#{$ns}-depth-chart {
  .canvas {
    display: block;
    cursor: crosshair;
    width: 100%;
    height: 100%;
  }

  &__toolbar {
    display: flex;

    button:disabled {
      opacity: 0.5;
    }
  }
}
