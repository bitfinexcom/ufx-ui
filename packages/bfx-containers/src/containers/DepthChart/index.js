import DepthChart from './DepthChart.container'
import DepthChartToolbar from './Toolbar.container'

export {
  DepthChart,
  DepthChartToolbar,
}

export default DepthChart
