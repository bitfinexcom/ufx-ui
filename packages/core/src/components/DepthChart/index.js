// eslint-disable-next-line import/no-named-as-default
import DepthChart from './DepthChart'
import DepthChartToolbar from './DepthChart.Toolbar'

export {
  DepthChartToolbar,
  DepthChart,
}

export default DepthChart
