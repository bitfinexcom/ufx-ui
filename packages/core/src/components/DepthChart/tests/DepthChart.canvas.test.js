import DepthChartCanvas from '../DepthChart.canvas'
import * as storyData from '../stories/DepthChart.stories_data'

const tests = describe('DepthChartCanvas', () => {
  let chart

  beforeEach(() => {
    document.write('<canvas id="depth-chart" style="width: 600px; height: 300px;"></canvas>')
    chart = new DepthChartCanvas({
      ref: document.querySelector('#depth-chart'),
      width: 600,
      height: 300,
      bidsColor: '#0f0',
      asksColor: '#f00',
      fontColor: '#8a9ba8',
      axisColor: '#8a9ba8',
      initialZoom: 50,
      onScroll: () => {},
      updateZoom: () => {},
    })
  })

  afterEach(() => {
    document.write('')
  })

  describe('initialize', () => {
    it('should initialize the canvas context correctly', () => {
      expect(chart.canvas).toBeTruthy()
      expect(chart.context.canvas).toBe(chart.canvas)
    })

    it('should initialize the dimensions correctly', () => {
      expect(chart.width).toBe(600)
      expect(chart.height).toBe(300)
    })
  })

  describe('render', () => {
    it('should clear the canvas', () => {
      chart.render({
        bids: storyData.bids,
        tBids: storyData.tBids,
        asks: storyData.asks,
        tAsks: storyData.tAsks,
        pBids: storyData.pBids,
        pAsks: storyData.pAsks,
        zoom: 25,
      })
      expect(chart.context.clearRect).toHaveBeenCalled()
    })

    it('should set midPrice correctly', () => {
      chart.render({
        bids: storyData.bids,
        tBids: storyData.tBids,
        asks: storyData.asks,
        tAsks: storyData.tAsks,
        pBids: storyData.pBids,
        pAsks: storyData.pAsks,
        zoom: 25,
      })
      expect(chart.midPrice).toBe(19275.5)
    })

    it('should set bids and asks correctly', () => {
      chart.render({
        bids: storyData.bids,
        tBids: storyData.tBids,
        asks: storyData.asks,
        tAsks: storyData.tAsks,
        pBids: storyData.pBids,
        pAsks: storyData.pAsks,
      })

      expect(chart.bids).toMatchSnapshot()
      expect(chart.asks).toMatchSnapshot()
    })

    it('should have sorted bids and asks by price', () => {
      chart.render({
        bids: storyData.bids,
        tBids: storyData.tBids,
        asks: storyData.asks,
        tAsks: storyData.tAsks,
        pBids: storyData.pBids,
        pAsks: storyData.pAsks,
        zoom: 50,
      })

      expect(chart.bids[0].price > chart.bids[1].price).toBe(false)
      expect(chart.bids[4].price > chart.bids[3].price).toBe(true)
      expect(chart.asks[1].price > chart.asks[2].price).toBe(false)
      expect(chart.asks[8].price > chart.asks[5].price).toBe(true)
    })

    it('should have set yTotal correctly', () => {
      chart.render({
        bids: storyData.bids,
        tBids: storyData.tBids,
        asks: storyData.asks,
        tAsks: storyData.tAsks,
        pBids: storyData.pBids,
        pAsks: storyData.pAsks,
        zoom: 100,
      })

      expect(chart.yTotal).toBe(Math.abs(storyData.tAsks.total))
    })

    it('should set x coordinates correctly', () => {
      chart.render({
        bids: storyData.bids,
        tBids: storyData.tBids,
        asks: storyData.asks,
        tAsks: storyData.tAsks,
        pBids: storyData.pBids,
        pAsks: storyData.pAsks,
        zoom: 100,
      })

      expect(chart.asks[6].x0).toBe(chart.getX(storyData.pAsks[6]))
      expect(chart.asks[6].x1).toBe(chart.getX(storyData.pAsks[7]))
    })

    it('should set y coordinate correctly', () => {
      chart.render({
        bids: storyData.bids,
        tBids: storyData.tBids,
        asks: storyData.asks,
        tAsks: storyData.tAsks,
        pBids: storyData.pBids,
        pAsks: storyData.pAsks,
        zoom: 25,
      })

      expect(chart.bids[4].y).toBe(chart.getY(chart.bids[4].total))
    })

    it('should call draw methods', () => {
      chart.drawBidsLine = jest.fn()
      chart.drawBidsGradient = jest.fn()
      chart.drawAsksLine = jest.fn()
      chart.drawAsksGradient = jest.fn()
      chart.drawYAxis = jest.fn()
      chart.drawXAxis = jest.fn()
      chart.drawXAxisTicks = jest.fn()
      chart.drawStaticTicks = jest.fn()
      chart.drawYAxisTicks = jest.fn()
      chart.drawMouseHover = jest.fn()

      chart.render({
        bids: storyData.bids,
        tBids: storyData.tBids,
        asks: storyData.asks,
        tAsks: storyData.tAsks,
        pBids: storyData.pBids,
        pAsks: storyData.pAsks,
      })

      expect(chart.drawBidsLine).toHaveBeenCalled()
      expect(chart.drawBidsGradient).toHaveBeenCalled()
      expect(chart.drawAsksLine).toHaveBeenCalled()
      expect(chart.drawAsksGradient).toHaveBeenCalled()
      expect(chart.drawYAxis).toHaveBeenCalled()
      expect(chart.drawXAxis).toHaveBeenCalled()
      expect(chart.drawXAxisTicks).toHaveBeenCalled()
      expect(chart.drawStaticTicks).toHaveBeenCalled()
      expect(chart.drawYAxisTicks).toHaveBeenCalled()
      expect(chart.drawMouseHover).toHaveBeenCalled()
    })
  })
})

export default tests
