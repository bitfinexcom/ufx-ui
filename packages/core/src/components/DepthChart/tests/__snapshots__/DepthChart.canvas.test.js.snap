// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DepthChartCanvas render should set bids and asks correctly 1`] = `
Array [
  Object {
    "amount": 2.0008,
    "difference": 14.03682309,
    "price": 19264,
    "total": 14.03682309,
    "x0": 23,
    "x1": 23,
    "y": 182,
  },
  Object {
    "amount": 0.6987225,
    "difference": -2.0008,
    "price": 19265,
    "total": 12.03602309,
    "x0": 23,
    "x1": 46.173913043478265,
    "y": 195,
  },
  Object {
    "amount": 1.1962,
    "difference": -0.6987225000000006,
    "price": 19266,
    "total": 11.33730059,
    "x0": 46.173913043478265,
    "x1": 70.34782608695653,
    "y": 199,
  },
  Object {
    "amount": 0.255,
    "difference": -1.1961999999999993,
    "price": 19267,
    "total": 10.14110059,
    "x0": 70.34782608695653,
    "x1": 94.52173913043478,
    "y": 207,
  },
  Object {
    "amount": 2.94979285,
    "difference": -0.2550000000000008,
    "price": 19268,
    "total": 9.88610059,
    "x0": 94.52173913043478,
    "x1": 118.69565217391305,
    "y": 209,
  },
  Object {
    "amount": 0.9963,
    "difference": -2.9497928499999997,
    "price": 19269,
    "total": 6.93630774,
    "x0": 118.69565217391305,
    "x1": 142.8695652173913,
    "y": 228,
  },
  Object {
    "amount": 1.501,
    "difference": -0.9962999999999997,
    "price": 19270,
    "total": 5.94000774,
    "x0": 142.8695652173913,
    "x1": 167.04347826086956,
    "y": 234,
  },
  Object {
    "amount": 0.0008,
    "difference": -1.5010000000000003,
    "price": 19271,
    "total": 4.43900774,
    "x0": 167.04347826086956,
    "x1": 191.21739130434784,
    "y": 244,
  },
  Object {
    "amount": 0.0008,
    "difference": -0.0007999999999999119,
    "price": 19272,
    "total": 4.43820774,
    "x0": 191.21739130434784,
    "x1": 215.3913043478261,
    "y": 244,
  },
  Object {
    "amount": 0.001,
    "difference": -0.0007999999999999119,
    "price": 19273,
    "total": 4.43740774,
    "x0": 215.3913043478261,
    "x1": 239.56521739130434,
    "y": 244,
  },
  Object {
    "amount": 0.001,
    "difference": -0.001000000000000334,
    "price": 19274,
    "total": 4.43640774,
    "x0": 239.56521739130434,
    "x1": 263.7391304347826,
    "y": 244,
  },
  Object {
    "amount": 4.43540774,
    "difference": -0.001000000000000334,
    "price": 19275,
    "total": 4.43540774,
    "x0": 263.7391304347826,
    "x1": 287.9130434782609,
    "y": 244,
  },
]
`;

exports[`DepthChartCanvas render should set bids and asks correctly 2`] = `
Array [
  Object {
    "amount": -26.36653064,
    "difference": -0.10000000000000142,
    "price": 19276,
    "total": 26.36653064,
    "x0": 312.0869565217391,
    "x1": 336.2608695652174,
    "y": 101,
  },
  Object {
    "amount": -0.1,
    "difference": -0.8399999999999999,
    "price": 19277,
    "total": 26.466530640000002,
    "x0": 336.2608695652174,
    "x1": 360.4347826086957,
    "y": 101,
  },
  Object {
    "amount": -0.84,
    "difference": -3.3348310199999993,
    "price": 19278,
    "total": 27.306530640000002,
    "x0": 360.4347826086957,
    "x1": 384.60869565217394,
    "y": 95,
  },
  Object {
    "amount": -3.33483102,
    "difference": -0.28889251000000016,
    "price": 19279,
    "total": 30.64136166,
    "x0": 384.60869565217394,
    "x1": 408.7826086956522,
    "y": 74,
  },
  Object {
    "amount": -0.28889251,
    "difference": -0.5875423700000013,
    "price": 19280,
    "total": 30.93025417,
    "x0": 408.7826086956522,
    "x1": 432.9565217391304,
    "y": 72,
  },
  Object {
    "amount": -0.58754237,
    "difference": -1.202272500000003,
    "price": 19281,
    "total": 31.517796540000003,
    "x0": 432.9565217391304,
    "x1": 457.1304347826087,
    "y": 68,
  },
  Object {
    "amount": -1.2022725,
    "difference": -2.3439445300000017,
    "price": 19282,
    "total": 32.720069040000006,
    "x0": 457.1304347826087,
    "x1": 481.30434782608694,
    "y": 60,
  },
  Object {
    "amount": -2.34394453,
    "difference": -0.6293270999999976,
    "price": 19283,
    "total": 35.06401357000001,
    "x0": 481.30434782608694,
    "x1": 505.4782608695652,
    "y": 45,
  },
  Object {
    "amount": -0.6293271,
    "difference": -1.6200352500000008,
    "price": 19284,
    "total": 35.693340670000005,
    "x0": 505.4782608695652,
    "x1": 529.6521739130435,
    "y": 41,
  },
  Object {
    "amount": -1.62003525,
    "difference": -2.643099999999997,
    "price": 19285,
    "total": 37.313375920000006,
    "x0": 529.6521739130435,
    "x1": 553.8260869565217,
    "y": 30,
  },
  Object {
    "amount": -2.6431,
    "difference": -0.9247999999999976,
    "price": 19286,
    "total": 39.95647592,
    "x0": 553.8260869565217,
    "x1": 577,
    "y": 13,
  },
  Object {
    "amount": -0.9248,
    "difference": 40.88127592,
    "price": 19287,
    "total": 40.88127592,
    "x0": 577,
    "x1": 577,
    "y": 7,
  },
]
`;
